File: dart/src/n04_observer/event_manager.dart
import 'dart:io';

import 'event_listener.dart';
import 'event_type.dart';

class EventManager {
    final Map<EventType, List<EventListener>> _listeners = {};

    void subscribe(final EventType eventType, final EventListener listener) {
        this._listeners.putIfAbsent(eventType, () => List.empty(growable: true));
        this._listeners[eventType]!.add(listener);
    }

    void unsubscribe(final EventType eventType, final EventListener listener) {
        this._listeners[eventType]!.remove(listener);
    }

    void notify(final EventType eventType, final File file) {
        if (this._listeners.containsKey(eventType)) {
            for (final EventListener listener in this._listeners[eventType]!) {
                listener.update(file);
            }
        }
    }
}

File: java/src/n04_observer/EventManager.java
package n04_observer;

import java.io.File;
import java.util.*;

public class EventManager {
    private Map<EventType, List<EventListener>> listeners;

    public EventManager() {
        this.listeners = new HashMap<>();
    }

    public void subscribe(EventType eventType, EventListener listener) {
        this.listeners.putIfAbsent(eventType, new ArrayList<>());
        this.listeners.get(eventType).add(listener);
    }

    public void unsubscribe(EventType eventType, EventListener listener) {
        this.listeners.get(eventType).remove(listener);
    }

    public void notify(EventType eventType, File file) {
        if (this.listeners.containsKey(eventType)) {
            for (EventListener listener : this.listeners.get(eventType)) {
                listener.update(file);
            }
        }
    }
}
